keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1475,225 +1382,400 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    MatchAttribute *string
 
-    // Requests describes all resources that are needed from the driver.
-    Requests []ResourceRequest
+    // Future extension, not part of the current design:
+    // A CEL expression which compares different devices and returns
+    // true if they match.
+    //
+    // Because it would be part of a one-of, old schedulers will not
+    // accidentally ignore this additional, for them unknown match
+    // criteria.
+    //
+    // matcher string
 }
 
-// ResourceRequest is a request for resources from one particular driver.
-type ResourceRequest struct {
-    // VendorParameters are arbitrary setup parameters for the requested
-    // resource. They are ignored while allocating a claim.
-    VendorParameters runtime.Object
+type ClaimConfigurationParameters struct {
+    // The configuration applies to devices in these requests.
+    //
+    // If empty, the configuration applies to all devices in the claim.
+    //
+    // +optional
+    RequestNames []string
 
-    ResourceRequestModel
+    ConfigurationParameters // inline
 }
 
-// ResourceRequestModel must have one and only one field set.
-type ResourceRequestModel struct {
-    // NamedResources describes a request for resources with the named resources model.
-    NamedResources *NamedResourcesRequest
+// ConfigurationParameters must have one and only one field set. It gets embedded
+// inline in some other structs which have other fields, so field names must
+// not conflict with those.
+type ConfigurationParameters struct {
+    Opaque *OpaqueConfigurationParameters
 }
 
-// NamedResourcesRequest is used in ResourceRequestModel.
-type NamedResourcesRequest struct {
-    // Selector is a CEL expression which must evaluate to true if a
-    // resource instance is suitable. The language is as defined in
-    // https://kubernetes.io/docs/reference/using-api/cel/
+// OpaqueConfigurationParameters contains configuration parameters for a driver.
+type OpaqueConfigurationParameters struct {
+    // DriverName is used to determine which kubelet plugin needs
+    // to be passed these configuration parameters.
+    //
+    // An admission webhook provided by the driver developer could use this
+    // to decide whether it needs to validate them.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
+    DriverName string
+
+    // Parameters can contain arbitrary data. It is the responsibility of
+    // the driver developer to handle validation and versioning. Typically this
+    // includes self-identification and a version ("kind" + "apiVersion" for
+    // Kubernetes types), with conversion between different versions.
+    Parameters runtime.RawExtension
+}
+```
+
+The `DeviceClassName` field may be left empty. The request itself can specify which
+driver needs to provide devices. There are some corner cases:
+- Empty `claim.requests`: this is a "null request" which can be satisfied without
```

Comment:
Why is ths useful?  Again, don't make nothing mean something.  How would you add that hypothetical non-device resource if this field being empty means "null" ?

Reply:
Now that we agree on requests[*].device, this COULD be empty, but I still don't see why we want to allow it.